Allow dates before 1600 to be valid

markstory opened the pull request:
Expand the valid date range to start at 0001 instead of 1600. This makes validating historical dates simpler as they are now valid.

Fixes #13441

markstory committed:
Allow dates before 1600 to be valid

Expand the valid date range to start at 0001 instead of 1600. This makes
validating historical dates simpler as they are now valid.

Fixes #13441

othercorey commented on tests/TestCase/Validation/ValidationTest.php:
```diff
@@ -1439,7 +1461,6 @@ public function testDateY()
         $this->assertFalse(Validation::date('20009', ['y']));
         $this->assertFalse(Validation::date(' 2012', ['y']));
         $this->assertFalse(Validation::date('3000', ['y']));
-        $this->assertFalse(Validation::date('1799', ['y']));
```
Should we replace this with an equivalent (although weird) somewhere? 31-12-0000

Is simply '0' or '0000' tested somewhere?

markstory replied:
The year 0 is currently not supported. From what I could understand from the gregorian calendar wikipedia's pages there isn't a year 0.

othercorey replied:
I meant technically that's the first invalid date.

othercorey commented on tests/TestCase/Validation/ValidationTest.php:
```diff
@@ -1429,6 +1450,7 @@ public function testDateYmNumeric()
      */
     public function testDateY()
     {
+        $this->assertTrue(Validation::date('0001', ['y']));
```
Are 3-digit years now supported?

markstory replied:
They shouldn't be. I can add a test for that though.

dereuromark replied:
That also means 2 and 1 digit years then, right?

markstory replied:
Supporting one, and two digit years gets hard as the 2 digit years are ambiguous with the modern day two digit format (which is a disaster, but still in use). Three digit years should be possible, but I'm not convinced the added complexity is justified given how infrequently these dates will be used.

othercorey replied:
I'm not convinced it's necessary either.  I think it'd be nice to have a failure test case.

markstory committed:
Don't accept 0000 or three digit years

Adding support for three digit years is possible, but I'm not convinced
that the added complexity is worth the trouble. The current patterns are
already very complex and date <1000 should be infrequently used.

markstory committed:
Extract repeated leap year patterns.
